qiskit/circuit/library/template_circuits/toffoli/template_4a_3.py

```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Template 4a_3:
+.. parsed-literal::
+    q_0: ───────■─────────■──
+                │         │
+    q_1: ──■────■────■────■──
+           │  ┌─┴─┐  │  ┌─┴─┐
+    q_2: ──┼──┤ X ├──┼──┤ X ├
+         ┌─┴─┐└───┘┌─┴─┐└───┘
+    q_3: ┤ X ├─────┤ X ├─────
+         └───┘     └───┘
+"""
```

Comment:
This is more of a general question about how these templates have been selected. I see in the paper you reference there is only one template 4(a), but here you generate `template_4a_1` through `template_4a_4`. Are these necessary? For example this template above I think can be derived from `template_4a_2` which is below. It is basically "rotating" the last gate to be first, then swapping wire q_2 with wire q_3.

```
"""
Template 4a_2:
.. parsed-literal::
    q_0: ──■─────────■───────
           │         │
    q_1: ──■────■────■────■──
           │  ┌─┴─┐  │  ┌─┴─┐
    q_2: ──┼──┤ X ├──┼──┤ X ├
         ┌─┴─┐└───┘┌─┴─┐└───┘
    q_3: ┤ X ├─────┤ X ├─────
         └───┘     └───┘
"""
```

I *think* the algorithm you have implemented can already detect this, no? i.e. if G1.G2.G3.G4 is a template, you don't *also* have to add G4.G1.G2.G3 as a template. Or do you?

The reason I think this is already supported by the algorithm is that when I use this template
```
                              
q_0: ──■─────────■─────────■──
     ┌─┴─┐     ┌─┴─┐       │  
q_1: ┤ X ├──■──┤ X ├──■────┼──
     └───┘┌─┴─┐└───┘┌─┴─┐┌─┴─┐
q_2: ─────┤ X ├─────┤ X ├┤ X ├
          └───┘     └───┘└───┘
```

on 
```
                    
q_0: ───────■───────
          ┌─┴─┐     
q_1: ──■──┤ X ├──■──
     ┌─┴─┐└───┘┌─┴─┐
q_2: ┤ X ├─────┤ X ├
     └───┘     └───┘
```

I get 

```
               
q_0: ──■────■──
       │  ┌─┴─┐
q_1: ──┼──┤ X ├
     ┌─┴─┐└───┘
q_2: ┤ X ├─────
     └───┘     
```

so even though the middle part matched it was able to reduce it. But maybe something else is happening in this case.

Reply:
You are right template_4a_3 is not necessary, I did a mistake when adding it. I'll remove it.